javaslang/src/main/java/javaslang/collection/Collections.java

```diff
@@ -62,16 +62,20 @@ static boolean areEqual(Iterable<?> iterable1, Iterable<?> iterable2) {
     static <T, C, R extends Iterable<T>> Map<C, R> groupBy(Traversable<T> source, Function<? super T, ? extends C> classifier, Function<? super Iterable<T>, R> mapper) {
         Objects.requireNonNull(classifier, "classifier is null");
         Objects.requireNonNull(mapper, "mapper is null");
-        final java.util.Map<C, Collection<T>> mutableResults = new java.util.LinkedHashMap<>(source.isTraversableAgain() ? source.size() : 16);
-        for (T value : source) {
-            final C key = classifier.apply(value);
-            mutableResults.computeIfAbsent(key, k -> new ArrayList<>()).add(value);
-        }
         Map<C, R> results = LinkedHashMap.empty();
-        for (java.util.Map.Entry<C, Collection<T>> entry : mutableResults.entrySet()) {
+        for (java.util.Map.Entry<? extends C, Collection<T>> entry : groupBy(source, classifier)) {
             results = results.put(entry.getKey(), mapper.apply(entry.getValue()));
         }
         return results;
+
+    }
+    private static <T, C> java.util.Set<java.util.Map.Entry<C, Collection<T>>> groupBy(Traversable<T> source, Function<? super T, ? extends C> classifier) {
```

Comment:
better separation: create a mutable one and convert it to persistent in separate steps